### PR TITLE
vo_dmabuf_wayland: support wp_viewport_set_source

### PR DESCRIPTION
**Pull request by boxerab**
the viewport source is adjusted whenever the source rectangle changes

### CODE REVIEW

File: video/out/vo_dmabuf_wayland.c [hunk 1]
```diff
@@ -191,6 +215,15 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
             p->want_reset = false;
     }
 
+    // 2. adjust viewport source if needed
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_osd_res osd;
+    mp_get_src_dst_rects(NULL, vo->opts, vo->driver->caps, vo->params,
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:222]: You should pass a log here.

Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c:222]: I intentionally pass null because I don't want logging for each frame drawn

Comment (Dudemanguy) [reply video/out/vo_dmabuf_wayland.c:222]: Oh that's a fair point. Just out of curiosity, is there a way to rework this check so it doesn't have to run on every frame? I guess it's not a big deal.

Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c:222]: as image can change dimensions at any point in container (for example with mpeg ts) I think it needs to run every frame.

**Comment by boxerab** [comment]
@Dudemanguy do you have any other issues with this PR ?

**Comment by Dudemanguy** [comment]
Panning still doesn't work correctly for me. The behavior is different from before though. Below is panning right. Trying to pan left shrinks the window.

2023-01-09-163742_grim

**Comment by boxerab** [comment]
@Dudemanguy pan should be working correctly now.

**Comment by Dudemanguy** [comment]
Nice, seems good. Mind squashing this into one commit? (`vo_dmabuf_wayland: support panscan and panning`)

**Comment by boxerab** [comment]
> Nice, seems good. Mind squashing this into one commit? (`vo_dmabuf_wayland: support panscan and panning`)

done

File: video/out/vo_dmabuf_wayland.c [hunk 2]
```diff
@@ -150,27 +153,78 @@ static bool drmprime_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_ent
 }
 #endif
 
+static void set_viewport_source(struct vo *vo, struct mp_rect src) {
+    struct priv *p = vo->priv;
+    struct vo_wayland_state *wl = vo->wl;
+
+    if (wl->video_viewport) {
+        if ( !mp_rect_equals(&p->src,&src)) {
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:161]:
```suggestion
    if (wl->video_viewport && !mp_rect_equals(&p->src, &src)) {
```

File: video/out/vo_dmabuf_wayland.c [hunk 3]
```diff
@@ -150,27 +153,78 @@ static bool drmprime_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_ent
...
+
+            // 1. update viewport source
+            wp_viewport_set_source(wl->video_viewport, src.x0 << 8,
+                                    src.y0 << 8, mp_rect_w(src) << 8,
+                                    mp_rect_h(src) << 8);
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:166]:
```suggestion
                                   src.y0 << 8, mp_rect_w(src) << 8,
                                   mp_rect_h(src) << 8);
```

File: video/out/vo_dmabuf_wayland.c [hunk 4]
```diff
@@ -150,27 +153,78 @@ static bool drmprime_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_ent
...
+    }
+}
+
+static void set_video_surface(struct vo_wayland_state *wl, struct mp_rect dst){
+    if (wl->video_viewport)
+        wp_viewport_set_destination(wl->video_viewport, mp_rect_w(dst), mp_rect_h(dst));
+    wl_subsurface_set_position(wl->video_subsurface, dst.x0, dst.y0);
+}
+
+static void set_video_surface_vo(struct vo *vo){
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:182]: It feels like `set_video_surface_vo` and `set_video_surface` should just be combined into `resize`. It's conceptually similar and would reduce some of the code duplication.

**Review by Dudemanguy — CHANGES_REQUESTED** [review]

File: video/out/vo_dmabuf_wayland.c (hunk 2 above)
Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c]: fixed

File: video/out/vo_dmabuf_wayland.c (hunk 4 above)
Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c]: good idea - fixed now.

File: video/out/vo_dmabuf_wayland.c (hunk 3 above)
Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c]: fixed

File: video/out/vo_dmabuf_wayland.c [hunk 5]
```diff
@@ -150,27 +153,56 @@ static bool drmprime_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_ent
 }
 #endif
 
+static void set_viewport_source(struct vo *vo, struct mp_rect src) {
+    struct priv *p = vo->priv;
+    struct vo_wayland_state *wl = vo->wl;
+
+    if (wl->video_viewport && !mp_rect_equals(&p->src,&src)) {
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:160]:
```suggestion
    if (wl->video_viewport && !mp_rect_equals(&p->src, &src)) {
```

File: video/out/vo_dmabuf_wayland.c [hunk 6]
```diff
@@ -376,6 +421,10 @@ static int preinit(struct vo *vo)
     };
     ra_hwdec_ctx_init(&p->hwdec_ctx, vo->hwdec_devs, NULL, true);
 
+    p->src.x0 = 0;
+    p->src.y0 = 0;
+    p->src.x1 = 0;
+    p->src.y1 = 0;
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:427]:
```suggestion
    p->src = (struct mp_rect){0, 0, 0, 0};
```

**Review by Dudemanguy — COMMENTED** [review]
LGTM other than those two small things.

File: video/out/vo_dmabuf_wayland.c (hunk 6 above)
Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c]: fixed

**Review by Dudemanguy — APPROVED** [review]

File: video/out/vo_dmabuf_wayland.c [hunk 7]
```diff
@@ -191,6 +223,15 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
...
     }
 
+    // 2. adjust viewport source if needed
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_osd_res osd;
+    mp_get_src_dst_rects(NULL, vo->opts, vo->driver->caps, vo->params,
+                         vo->dwidth, vo->dheight, vo->monitor_par,
+                         &src, &dst, &osd);
+    set_viewport_source(vo, src);
```
Comment (Dudemanguy) [inline video/out/vo_dmabuf_wayland.c:233]: Sorry, I know I just approved this, but I just realized that you can actually just move this to resize right after you're doing setting the subsurface position. No need to run this every frame.

**Review by Dudemanguy — CHANGES_REQUESTED** [review]

File: video/out/vo_dmabuf_wayland.c (hunk 7 above)
Comment (boxerab) [reply video/out/vo_dmabuf_wayland.c:233]: Thanks, but we do need to run it each frame - there can be a frame transition without a resize.

Comment (Dudemanguy) [reply video/out/vo_dmabuf_wayland.c:233]: What do you mean? If the panscan/pan change, `VOCTRL_SET_PANSCAN` will handle it.

**Comment by boxerab** [comment]
[reschange.txt](https://github.com/mpv-player/mpv/files/10394220/reschange.txt)

**Review by Dudemanguy — APPROVED** [review]

**Pull request merged**